### PR TITLE
SDKS-1832: Split Filter config

### PR DESCRIPTION
**Pull request by javrudsky**
# Android SDK

## What did you accomplish?
- Added SplitFilters to split client config
- Added unit tests

### CODE REVIEW

File: src/main/java/io/split/android/client/SyncConfig.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        private final SplitValidator mSplitValidator = new SplitValidatorImpl();
+
+        public SyncConfig build() {
+            List<SplitFilter> filters = new ArrayList<>();
+            for (SplitFilter filter : mBuilderFilters) {
+
+                List<String> values = filter.getValues();
+                switch (filter.getType()) {
+                    case BY_NAME:
+                        if (values.size() > MAX_BY_NAME_VALUES) {
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/SyncConfig.java:46]: You should move the limit validation after invalid and duplicated values are removed, and do it only once when all values are grouped by type.

File: src/main/java/io/split/android/client/SyncConfig.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                            String message = "Error: 400 different split names can be specified at most. You passed " +
+                                    values.size()
+                                    + ". Please consider reducing the amount or using prefixes to target specific groups of splits.";
+                            throw new IllegalArgumentException(message);
+                        }
+                        break;
+                    case BY_PREFIX:
+                        if (values.size() > MAX_BY_PREFIX_VALUES) {
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/SyncConfig.java:54]: maybe you don't need the switch-case sentence. Instead you can define a `getMaxSize` method in SplitFilter.Type enum, and use it somehow as follows:

```
if (values.size() > filter.getType().getMaxSize()) { String message = "Error: " + filter.getType().getMaxSize() + ...
```

File: src/main/java/io/split/android/client/SyncConfig.java [hunk 3]
```diff
@@ -37,51 +35,35 @@ public static Builder builder() {
...
-                                    values.size() +
-                                    "Please consider using a lower number of prefixes and/or filtering by split name as well.";
-
-                            throw new IllegalArgumentException(message);
-                        }
-                        break;
-                    default:
-                        Logger.e("Invalid Split filter!");
+                List<String> groupedValues = groupedFilters.get(filter.getType());
+                if (groupedValues == null) {
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/SyncConfig.java:41]: you can init the map lists outside the for each, as follows:

```
            Map<SplitFilter.Type, List<String>> groupedFilters = new HashMap<>();
            for(SplitFilter.Type type: SplitFilter.Type.values()) {
                groupedFilters.put(type, new ArrayList<>(););
            }
            for (SplitFilter filter : mBuilderFilters) {
                List<String> groupedValues = groupedFilters.get(filter.getType());

                List<String> values = filter.getValues();
                for (String value : values) {
                    if (mSplitValidator.validateName(value) != null) {
                        Logger.w(String.format("Warning: Malformed %s value. Filter ignored: %s", filter.getType().toString(), value));
                    } else {
                        groupedValues.add(value);
                    }
                }
            }
```

File: src/main/java/io/split/android/client/SyncConfig.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+                List<String> values = filter.getValues();
+                for (String value : values) {
+                    if (mSplitValidator.validateName(value) != null) {
+                        Logger.w(String.format("Warning: Malformed %s value. Filter ignored: %s", filter.getType().toString(), value));
+                    } else {
+                        groupedValues.add(value);
+                    }
+                }
+                groupedFilters.put(filter.getType(), groupedValues);
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/SyncConfig.java:53]: Dont't forget to move this line inside the if statement ;)

Comment (javrudsky) [reply src/main/java/io/split/android/client/SyncConfig.java]: Great. Done Doc!

**Pull request merged**